nifi-nar-bundles/nifi-cdc/nifi-cdc-mysql-bundle/nifi-cdc-mysql-processors/src/main/java/org/apache/nifi/cdc/mysql/processors/CaptureChangeMySQL.java

```diff
@@ -967,13 +1026,19 @@ public void outputEvents(ProcessSession session, ComponentLog log) throws IOExce
                             CommitTransactionEventInfo commitTransactionEvent = useGtid
                                     ? new CommitTransactionEventInfo(currentDatabase, timestamp, currentGtidSet)
                                     : new CommitTransactionEventInfo(currentDatabase, timestamp, currentBinlogFile, currentBinlogPosition);
-                            currentSequenceId.set(commitEventWriter.writeEvent(currentSession, transitUri, commitTransactionEvent, currentSequenceId.get(), REL_SUCCESS));
+                            currentEventInfo = commitTransactionEvent;
+                            currentEventWriter = commitEventWriter;
+                            currentSequenceId.set(commitEventWriter.writeEvent(currentSession, transitUri, commitTransactionEvent, currentSequenceId.get(), REL_SUCCESS, eventWriterConfiguration));
                         }
+
                         //update inTransaction value to state
                         inTransaction = false;
                         updateState(session);
-                        // Commit the NiFi session
-                        session.commitAsync();
+                        // If there is no FlowFile open, commit the session
+                        if (eventWriterConfiguration.getCurrentFlowFile() == null) {
+                            // Commit the NiFi session
+                            session.commitAsync();
```

Comment:
The process session is committed only after a transaction which is somewhat desirable. But because of this the N events per flowfile can hold back a flowfile with N events already in it until a commit comes. But the commit may never actually come.
Consider the following case:
- N=5
- 1st transaction inserts 4 files
- 2nd transaction inserts 2 files.
- When the event for the first insert of the 2nd transaction is received we finish up the current flowfile (has 4 events from the 1st transaction and 1 event from the 2nd transaction). However we don't commit the process session - we are not handling a commit event! - so the flowfile is held back.
- When the commit for the 2nd transaction is received we have only 1 event in the flowfile and it is still worked on (not null). So we don't commit the process session.
- A bunch of flowfiles can queue up this way until we reach a point where the commit comes when the flowfile has exactly N events or when we stop the processor.

Reply:
In that case we should call the property `Max Events Per FlowFile` and document that when the event queue is empty (about to finish a call to onTrigger), we will finish any open FlowFile and commit the session. @tpalfy  @exceptionfactory @mark-bathori thoughts?

Reply:
Yeah, I think that is more desirable anyway.